hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/capacity/QueueStateHelper.java

```diff
@@ -0,0 +1,96 @@
+package org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity;
+
+import org.apache.hadoop.thirdparty.com.google.common.collect.ImmutableSet;
+import org.apache.hadoop.yarn.api.records.QueueState;
+import org.apache.hadoop.yarn.exceptions.YarnException;
+
+import java.util.Set;
+
+/**
+ * Collects all logic that are handling queue state transitions.
+ */
+public final class QueueStateHelper {
+  private static final Set<QueueState> VALID_STATE_CONFIGURATIONS = ImmutableSet.of(
+      QueueState.RUNNING, QueueState.STOPPED);
+  private static final QueueState DEFAULT_STATE = QueueState.RUNNING;
+
+  private QueueStateHelper() {}
+
+  /**
+   * Sets the current state of the queue based on its previous state, its parent's state and its
+   * configured state.
+   * @param queue the queue whose state is set
+   */
+  public static void setQueueState(AbstractCSQueue queue) {
+    QueueState previousState = queue.getState();
+    QueueState configuredState = queue.getQueueContext().getConfiguration().getConfiguredState(
+        queue.getQueuePath());
+    QueueState parentState = (queue.getParent() == null) ? null : queue.getParent().getState();
+
+    // verify that we can not any value for State other than RUNNING/STOPPED
+    if (configuredState != null && !VALID_STATE_CONFIGURATIONS.contains(configuredState)) {
+      throw new IllegalArgumentException("Invalid queue state configuration."
+          + " We can only use RUNNING or STOPPED.");
+    }
+
+    if (previousState == null) {
+      initializeState(queue, configuredState, parentState);
+    } else {
+      reinitializeState(queue, previousState, configuredState);
+    }
+  }
+
+  private static void reinitializeState(
+      AbstractCSQueue queue, QueueState previousState, QueueState configuredState) {
+    // when we get a refreshQueue request from AdminService,
+    if (previousState == QueueState.RUNNING) {
+      if (configuredState == QueueState.STOPPED) {
+        queue.stopQueue();
+      }
+    } else {
+      if (configuredState == QueueState.RUNNING) {
+        try {
+          queue.activateQueue();
+        } catch (YarnException ex) {
+          throw new IllegalArgumentException(ex.getMessage());
+        }
+      }
+    }
+  }
+
+  private static void initializeState(
+      AbstractCSQueue queue, QueueState configuredState, QueueState parentState) {
+    QueueState currentState = configuredState == null ? DEFAULT_STATE : configuredState;
+
+    if (configuredState == QueueState.RUNNING && parentState != QueueState.RUNNING) {
+      throw new IllegalArgumentException(
+          "The parent queue:" + queue.getParent().getQueuePath()
+              + " cannot be STOPPED as the child queue:" + queue.getQueuePath()
+              + " is in RUNNING state.");
+    }
+
+    if (parentState != null && configuredState == null) {
+      currentState = parentState == QueueState.DRAINING ? QueueState.STOPPED : parentState;
+    }
+
+    queue.updateQueueState(currentState);
```

Comment:
Never mind, an exception will be thrown in that case.

Reply:
Correct, is 
``` 
if (configuredState == QueueState.RUNNING && parentState != null &&
        parentState != QueueState.RUNNING) {
      throw new IllegalArgumentException(
          "The parent queue:" + queue.getParent().getQueuePath()
              + " cannot be STOPPED as the child queue:" + queue.getQueuePath()
              + " is in RUNNING state.");
    }
```
sufficient here? Thanks for the resource, it will be useful for future refactors like this.

Reply:
Yes, I think it's correct now. Thanks for working this, the latest change looks good to me. Non-binding +1.